astroid/bases.py

```diff
@@ -436,7 +436,14 @@ def _infer_builtin_new(
         node_context = context.extra_context.get(caller.args[0])
         infer = caller.args[0].infer(context=node_context)
 
-        yield from (Instance(x) if x is not Uninferable else x for x in infer)  # type: ignore[misc,arg-type]
+        for inferred in infer:
+            if isinstance(inferred, type(Uninferable)):  # isinstance helps mypy
+                yield inferred
+            if isinstance(inferred, Instance):
+                raise InferenceError
```

Comment:
If I understand correctly a single instance of type ``Instance`` raise an InferanceError and void every value in infer ? Why don't we yield ``Uninferable`` instead ?

Reply:
Yes, it would prevent reaching any subsequent values (not prior values). We could go either way, but this seemed more consistent with the rest of the system to me.

Examples:
- #782
- https://github.com/PyCQA/astroid/blob/3621e2e7d68653d66cbf770e6dcb61ba541117f1/astroid/arguments.py#L176-L187
- pylint's `safe_infer_call_result`
- the various `try/except` around `infer_call_result` that catch InferenceError instead of checking for Uninferable: https://github.com/PyCQA/pylint/blob/01ea163c48d759f86d75dfb47e00796b7409abe1/pylint/checkers/classes/class_checker.py#L1184-L1187

Reply:
Thank you !